add watermark metrics

tomachine opened the pull request:
This solves #310 

Adding 4 new metrics: Disk allocation watermark high, low, flood_stage and disk threshold_enabled.
Metrics support both types of disk allocation watermarks - size in bytes and in percent (shown in label of the metrics).

lcaflc commented:
I add support on this feature. This is an highly needed metric for the monitoring of a production cluster. This allow a clean monitoring state of the ES cluster disk usage status without the need to create host dependent disk usage alarms which is painfull with multiple cluster using different watermark levels.

sysadmind commented on collector/cluster_settings.go:
```diff
@@ -21,22 +21,52 @@ import (
+	ValueWithUnit func(clusterSettings ClusterSettingsResponse) (float64, string)
+}
+
+func getUnitAndValue(value string) (float64, string) {
```
I don't see any examples where the values have bytes, only percentage. Do you know which version this changed? 

It is against the best practices to have units be a label. Units should always be normalized (i.e. bytes or seconds). I think it would be best to keep the percentage here as a gauge, but convert percentage to a float from 0-1. 

I'm also not sure about bringing in another dependency to do the unit conversion.

frenkye replied:
@sysadmind Hi, sorry for delay here. @tomachine prepared this for my issue. We aproach it this way, because this setting could be set both ways and this was sounds acceptable when he wrote this.

If you have other suggestion how to approach this, we are open to suggestions.

You can read about this in official docs and I think this was possible set to percent or absolute value in version 6. Link: https://www.elastic.co/guide/en/elasticsearch/reference/8.6/modules-cluster.html#disk-based-shard-allocation

> cluster.routing.allocation.disk.watermark.low logo cloud
    (Dynamic) Controls the low watermark for disk usage. It defaults to 85%, meaning that Elasticsearch will not allocate shards to nodes that have more than 85% disk used. It can alternatively be set to a ratio value, e.g., 0.85. It can also be set to an absolute byte value (like 500mb) to prevent Elasticsearch from allocating shards if less than the specified amount of space is available. This setting has no effect on the primary shards of newly-created indices but will prevent their replicas from being allocated.

sysadmind replied:
Reading the docs, I can see where this value may be set to a byte value. The docs also mention that the value could be a ratio like `0.8` so that should also be considered. 

Looking over this code again, I think overall it is probably pretty close to where it needs to be, however I'm not a fan of the added 3rd party dependency. The package does not seem to be widely used and it means this exporter has to depend on an outside party for another package.  I'm also not sure, based on the tests in that package that it would properly parse a value like `500mb` anyway.

rdegez commented:
Would love this feature too (at least percentage) so we can have dynamic alerting threshold based on configured wattermarks and not arbitrary static ones.

tomachine commented:
I rebased and removed the 3rd party dependency. Also solved the issue with value in bytes/ratio/percents watermark by two types of metrics - that will be collected according to type of watermark value.

sysadmind commented on collector/cluster_settings.go:
```diff
@@ -151,5 +208,106 @@ func (c *ClusterSettingsCollector) Update(ctx context.Context, ch chan<- prometh
 	return nil
 }
+
+func getValueInBytes(value string) float64 {
```
As a followup, this should return an error so that the metric is not reported and an error is logged when we fail to parse. There are also no tests for this function or the one below to make sure that no future changes break this functionality.
